homeassistant/components/ted5000/sensor.py

```diff
@@ -49,85 +165,148 @@ def setup_platform(
 
-    def update(self) -> None:
-        """Get the latest data from REST API."""
-        self._gateway.update()
+        if plan_type in (0, 2):
+            current_tier = 0
+        else:
+            current_tier = int(doc["LiveData"]["Utility"]["CurrentTier"]) + 1
+
+        if plan_type < 2:
+            current_tou = 0
+            current_tou_str = "Not Configured"
+        else:
+            current_tou = int(doc["LiveData"]["Utility"]["CurrentTOU"]) + 1
+            current_tou_str = doc["LiveData"]["Utility"]["CurrentTOUDescription"]
+
+        self.data_utility = {
+            ENTITY_MTUS: mtus,
+            ENTITY_RATE: current_rate / 100000,
+            ENTITY_DAYSLEFT: days_left,
+            ENTITY_PLANTYPE: plan_type_str[plan_type],
+            ENTITY_TIER: current_tier,
+            ENTITY_TOU: current_tou,
+            ENTITY_TOUDESC: current_tou_str,
+            ENTITY_CARBONRATE: carbon_rate / 100,
+            ENTITY_METERREAD: read_date,
+        }
 
+        return mtus
 
-class Ted5000Gateway:
-    """The class for handling the data retrieval."""
+    return 0
+
+
+class Ted5000Data:
+    """Collect data from the ted5000 gateway."""
 
     def __init__(self, url):
         """Initialize the data object."""
         self.url = url
-        self.data = {}
+        self.mtus = 0
 
     @Throttle(MIN_TIME_BETWEEN_UPDATES)
     def update(self):
-        """Get the latest data from the Ted5000 XML API."""
+        """Get the latest data from ted5000."""
+        # Update data
+        self.mtus = get_ted5000(self)
 
-        try:
-            request = requests.get(self.url, timeout=10)
-        except requests.exceptions.RequestException as err:
-            _LOGGER.error("No connection to endpoint: %s", err)
-        else:
-            doc = xmltodict.parse(request.text)
-            mtus = int(doc["LiveData"]["System"]["NumberMTU"])
 
-            for mtu in range(1, mtus + 1):
-                power = int(doc["LiveData"]["Power"]["MTU%d" % mtu]["PowerNow"])
-                voltage = int(doc["LiveData"]["Voltage"]["MTU%d" % mtu]["VoltageNow"])
+class Ted5000SensorEntity(SensorEntity):
+    """Representation of the current ted5000 data."""
 
-                self.data[mtu] = {
-                    POWER_WATT: power,
-                    ELECTRIC_POTENTIAL_VOLT: voltage / 10,
-                }
+    _attr_has_entity_name = True
+
+    def __init__(
+        self,
+        gateway,
+        mtu,
+        description: SensorEntityDescription,
+        name,
+    ):
+        """Initialize the sensor."""
+        self.entity_description = description
+        if mtu > 0:
+            self._attr_name = f"{name} MTU{mtu} {description.name}"
+        else:
+            self._attr_name = f"{name} Utility {description.name}"
+        self.gateway = gateway
+        self._mtu = mtu
+
+    def update(self):
+        """Get the latest data from Ted5000 and update the state."""
+        self.gateway.update()
```

Comment:
You should be aware that doing this will attempt to call the API and get the full data set for each sensor entity (could be 14 times).

I see the @throttle statement above (which I am not familiar with).

Is the strategy to use the rate limit to cause only the first request to go through?

This might be ok, but if so, I think you should add a comment to say that's what you are doing.


DataUpdateCoordinator resolves this as previously discussed.

Reply:
I don't think it would call it 14 times, this is called once every scan interval and is limited through the **@Throttle** to a minimum 10 seconds between recalls. Once called, it reads the entire XML once and extract data from it once and returns data once.

I'll add a comment

Reply:
But if you have 14 sensors then the update() function will be called 14 times.

Reply:
Possibly it is worth a bit of print() debugging to see if there are lots of calls being made.

Relying on the @throttle statement might be completely valid, or perhaps I've misunderstood something and it's completely fine. ;)

Reply:
Unfortunately, you're right as it calls it many times... I have absolutely no clue how to prevent this...